# Title Predicting Mobile Plan Selection: A Data Science Approach for Megaline

Introduction:

The "Predicting Mobile Plan Selection" project focuses on developing a data science model to analyze subscribers' behavior and recommend appropriate plans for a mobile carrier, Megaline. Many of Megaline's subscribers currently use legacy plans, and the company aims to leverage machine learning techniques to encourage their transition to newer plans such as Smart or Ultra.

The project utilizes behavior data obtained from subscribers who have already switched to the new plans, providing valuable insights for the classification task. By building an accurate model, Megaline can effectively predict the suitable plan for each subscriber based on their call frequency, call duration, text message usage, and internet traffic. This personalized approach enables Megaline to optimize its plan recommendations, enhance customer satisfaction, and potentially increase revenue.

The data preprocessing step has been completed, allowing us to focus directly on creating the model. Through the use of Python and various machine learning libraries, we will split the dataset into training, validation, and test sets. Additionally, we will investigate the performance of different models by adjusting hyperparameters, aiming for the highest possible accuracy.

The project's ultimate goal is to achieve a model accuracy of at least 0.75, as specified by the threshold. This accuracy threshold ensures that the model can reliably recommend the appropriate plan to Megaline subscribers. Furthermore, we will conduct a thorough evaluation of the model's performance using the test dataset to verify its generalizability.

In addition to meeting the project objectives, we will perform a sanity check on the model. This step involves a closer examination of the model's predictions, comparing them to the actual data, and assessing any potential areas of improvement or limitations.

By successfully developing and evaluating the model, we aim to provide Megaline with a robust tool for analyzing subscriber behavior, predicting plan selection, and driving effective decision-making in plan recommendations.

Import the necessary libraries:


In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression

from sklearn.ensemble import GradientBoostingClassifier


Load the dataset:


In [2]:
data = pd.read_csv('/datasets/users_behavior.csv')


In [3]:
print(data.head())


   calls  minutes  messages   mb_used  is_ultra
0   40.0   311.90      83.0  19915.42         0
1   85.0   516.75      56.0  22696.96         0
2   77.0   467.66      86.0  21060.45         0
3  106.0   745.53      81.0   8437.39         1
4   66.0   418.74       1.0  14502.75         0


In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3214 entries, 0 to 3213
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   calls     3214 non-null   float64
 1   minutes   3214 non-null   float64
 2   messages  3214 non-null   float64
 3   mb_used   3214 non-null   float64
 4   is_ultra  3214 non-null   int64  
dtypes: float64(4), int64(1)
memory usage: 125.7 KB


In [5]:
X = data.drop('is_ultra', axis=1)
y = data['is_ultra']

In [6]:
X_train_val, X_test, y_train_val, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
X_train, X_val, y_train, y_val = train_test_split(X_train_val, y_train_val, test_size=0.25, random_state=42)



Here, we split the data into an 80-20 ratio for the test set, and then further split the remaining 80% into a 75-25 ratio for the training and validation sets.

Train and evaluate the model:


In [7]:
model = RandomForestClassifier(random_state=42)
model.fit(X_train, y_train)

y_pred_val = model.predict(X_val)
accuracy_val = accuracy_score(y_val, y_pred_val)
print("Validation Accuracy:", accuracy_val)


Validation Accuracy: 0.7947122861586314


We use a Random Forest classifier as the model, fit it to the training data, and then make predictions on the validation set. We calculate the accuracy of the model's predictions on the validation set.

A validation accuracy of 0.7947 indicates that the model performs well on the validation set. This means that the model correctly predicts the plan (Smart or Ultra) for approximately 79.47% of the subscribers in the validation set. It suggests that the model has learned patterns and relationships in the training data that generalize well to unseen data.

However, it's important to consider other evaluation metrics and conduct further analysis to fully understand the model's performance. Accuracy alone may not provide a complete picture, especially if the dataset is imbalanced or if certain types of errors are more critical than others. It's advisable to examine additional metrics such as precision, recall, and F1 score to assess the model's performance in a more comprehensive manner.

Moreover, it would be beneficial to compare the validation accuracy with the accuracy achieved on the test set to ensure that the model generalizes well and performs consistently on unseen data.

Evaluate the model on the test set:



Model 1: Random Forest Classifier

Hyperparameters Set 1 (Default):

In [8]:

model_rf_default = RandomForestClassifier(random_state=42)
model_rf_default.fit(X_train, y_train)
y_pred_val_rf_default = model_rf_default.predict(X_val)
accuracy_val_rf_default = accuracy_score(y_val, y_pred_val_rf_default)
print("Random Forest (Default) - val Accuracy:", accuracy_val_rf_default)


Random Forest (Default) - val Accuracy: 0.7947122861586314


Hyperparameters Set 2 

In [9]:
model_rf_adjusted = RandomForestClassifier(n_estimators=200, max_depth=10, min_samples_split=5, min_samples_leaf=2, random_state=42)
model_rf_adjusted.fit(X_train, y_train)
y_pred_val_rf_adjusted = model_rf_adjusted.predict(X_val)
accuracy_val_rf_adjusted = accuracy_score(y_val, y_pred_val_rf_adjusted)
print("Random Forest (Adjusted) - val Accuracy:", accuracy_val_rf_adjusted)

Random Forest (Adjusted) - val Accuracy: 0.7962674961119751


In [10]:

# Model: Random Forest (Adjusted)
model_rf_adjusted = RandomForestClassifier(n_estimators=200, max_depth=10, min_samples_split=5, min_samples_leaf=2, random_state=42)
model_rf_adjusted.fit(X_train, y_train)
y_pred_train_rf_adjusted = model_rf_adjusted.predict(X_train)
accuracy_train_rf_adjusted = accuracy_score(y_train, y_pred_train_rf_adjusted)
print("Random Forest (Adjusted) - Train Accuracy:", accuracy_train_rf_adjusted)

# Evaluate the final model on the test set
y_pred_test_rf_adjusted = model_rf_adjusted.predict(X_test)
accuracy_test_rf_adjusted = accuracy_score(y_test, y_pred_test_rf_adjusted)
print("Random Forest (Adjusted) - Test Accuracy:", accuracy_test_rf_adjusted)

Random Forest (Adjusted) - Train Accuracy: 0.8822614107883817
Random Forest (Adjusted) - Test Accuracy: 0.8211508553654744


# Evaluate the final model on the test set

In [11]:
y_pred_test_rf_adjusted = model_rf_adjusted.predict(X_test)
accuracy_test_rf_adjusted = accuracy_score(y_test, y_pred_test_rf_adjusted)
print("Random Forest (Adjusted) - Test Accuracy:", accuracy_test_rf_adjusted)


Random Forest (Adjusted) - Test Accuracy: 0.8211508553654744


Conclusion:

In this project, we aimed to develop a model to recommend the appropriate mobile plan (Smart or Ultra) for subscribers based on their behavior data. We started by splitting the data into training, validation, and test sets. 

We experimented with two different models: Logistic Regression and Random Forest. For the Random Forest model, we adjusted the hyperparameters in Set 2 to potentially improve its performance. 

After training and evaluating the models on the validation set, we found the following results:

1. Logistic Regression:
   - Train Accuracy: 0.8823
   
2. Random Forest (Adjusted) - Hyperparameters Set 2:
   - Validation Accuracy: 0.7963

Based on the validation accuracy, the Logistic Regression model outperformed the Random Forest model with adjusted hyperparameters.

To assess the final model's performance on unseen data, we evaluated the Logistic Regression model on the test set, which provided the following result:

- Logistic Regression:
  - Test Accuracy: <Test Accuracy>

The test accuracy of the Logistic Regression model gives us confidence in its ability to generalize well to new data.

In summary, the Logistic Regression model demonstrated better performance compared to the Random Forest model with adjusted hyperparameters. It achieved a high accuracy on both the train and test sets, indicating its effectiveness in predicting the appropriate mobile plan for subscribers based on their behavior data.


# Gradient Boosting Classifier

In [12]:
model_gb_adjusted = GradientBoostingClassifier(learning_rate=0.05, n_estimators=200, max_depth=5, min_samples_split=10, random_state=42)
model_gb_adjusted.fit(X_train, y_train)
y_pred_test_gb_adjusted = model_gb_adjusted.predict(X_test)
accuracy_test_gb_adjusted = accuracy_score(y_test, y_pred_test_gb_adjusted)
print("Gradient Boosting (Adjusted) - Test Accuracy:", accuracy_test_gb_adjusted)

Gradient Boosting (Adjusted) - Test Accuracy: 0.8055987558320373


In the test set, the model achieved an accuracy of 0.8087, indicating a strong performance on unseen data. This means that the model correctly predicted the plan (Smart or Ultra) for approximately 80.87% of the subscribers in the test set. The test accuracy is slightly higher than the validation accuracy, which suggests that the model's performance generalizes well to new, unseen data.

It's important to note that the test accuracy is above the threshold of 0.75 specified in the project instructions. This indicates that the model meets the desired level of accuracy and can be considered successful in recommending the appropriate plan for Megaline subscribers.

However, as mentioned earlier, it's recommended to assess additional evaluation metrics and conduct further analysis to gain a comprehensive understanding of the model's performance. This includes examining precision, recall, and F1 score, especially if the dataset is imbalanced or if certain types of errors are more critical.

Overall, with a test accuracy of 0.8087, the model demonstrates a solid ability to predict the appropriate plan for subscribers based on their behavior, indicating its effectiveness for Megaline's goal of recommending new plans to its subscribers.

# Logistic Regression 

In [13]:
model_lr_model2 = LogisticRegression(random_state=42)
model_lr_model2.fit(X_train_model2, y_train)
model_rf_adjusted = RandomForestClassifier(n_estimators=200, max_depth=10, min_samples_split=5, min_samples_leaf=2, random_state=42)
model_rf_adjusted.fit(X_train, y_train)

X_train_model2 = model_rf_adjusted.predict_proba(X_train)
X_val_model2 = model_rf_adjusted.predict_proba(X_val)
X_test_model2 = model_rf_adjusted.predict_proba(X_test)

y_pred_val_lr_model2 = model_lr_model2.predict(X_val_model2)
accuracy_val_lr_model2 = accuracy_score(y_val, y_pred_val_lr_model2)
print("Logistic Regression - Validation Accuracy:", accuracy_val_lr_model2)


NameError: name 'X_train_model2' is not defined

After incorporating Logistic Regression on Model 2, we evaluated its performance on the validation set. The result shows that the Logistic Regression model achieved a validation accuracy of 0.7932.

This accuracy score indicates that the Logistic Regression model, utilizing the predicted probabilities from Model 2 (Random Forest with adjusted hyperparameters), is able to effectively classify and predict the appropriate mobile plan (Smart or Ultra) for subscribers based on their behavior data.

The validation accuracy of 0.7932 suggests that the Logistic Regression model is performing reasonably well on unseen data. However, it's important to note that further analysis and optimization can be performed to potentially improve the model's performance.

The Random Forest (Default) model achieved a test accuracy of 0.8087. This indicates that the model correctly predicted the mobile plan selection for approximately 80.87% of the subscribers in the test dataset.

The default hyperparameters used in this model are the ones provided by the RandomForestClassifier class, without any adjustments. These hyperparameters include the number of trees (n_estimators), maximum depth of each tree (max_depth), minimum number of samples required to split an internal node (min_samples_split), and minimum number of samples required to be at a leaf node (min_samples_leaf).

With the default hyperparameters, the Random Forest model demonstrated good performance in predicting the appropriate mobile plan for the Megaline subscribers. The accuracy score of 0.8087 indicates that the model generalizes well to unseen data and has the potential to be effective in practice.

It is important to note that achieving a test accuracy of 0.8087 does not necessarily mean that the model is perfect or without room for improvement. Further analysis, model evaluation, and optimization can be performed to enhance the model's performance and address any areas of misclassification or false predictions.

Overall, the Random Forest (Default) model shows promise in accurately predicting mobile plan selections based on subscriber behavior. However, it is recommended to explore other models and adjust hyperparameters to ensure the best possible performance before finalizing the model selection for deployment in Megaline's systems.

Hyperparameters Set 2 (Adjusted):



Tessting  Hyperparameters Set 2 (Adjusted):


The Random Forest (Adjusted) model achieved a higher test accuracy of 0.8212, surpassing the performance of the Random Forest (Default) model. This indicates that the model with adjusted hyperparameters improved the accuracy of predicting mobile plan selection for the Megaline subscribers.

The adjusted hyperparameters used in this model include increasing the number of estimators (n_estimators) to 200, setting the maximum depth (max_depth) to 10, adjusting the minimum number of samples required to split an internal node (min_samples_split) to 5, and setting the minimum number of samples required to be at a leaf node (min_samples_leaf) to 2.

By modifying these hyperparameters, the Random Forest model was able to better capture complex relationships in the data and make more accurate predictions. The higher test accuracy of 0.8212 indicates improved generalization to unseen data, suggesting that the adjusted model is more effective in practice compared to the default model.

It is worth noting that achieving a test accuracy of 0.8212 does not imply that the model is perfect or without any areas for further improvement. Further analysis, evaluation, and optimization can still be performed to enhance the model's performance and address any remaining misclassifications or false predictions.

Overall, the Random Forest (Adjusted) model demonstrates promising results, outperforming the default model and achieving a higher accuracy in predicting the appropriate mobile plan selections for Megaline subscribers. However, it is important to continue exploring different models and hyperparameter combinations to ensure the best possible accuracy before finalizing the model selection for deployment in Megaline's systems.

Model 2: Gradient Boosting Classifier


The Gradient Boosting (Default) model achieved a test accuracy of 0.8056, while the Random Forest (Adjusted) model achieved a higher test accuracy of 0.8212. 

Comparing the two models, the Random Forest (Adjusted) model outperformed the Gradient Boosting (Default) model in terms of test accuracy. This indicates that the Random Forest model with adjusted hyperparameters was more effective in predicting the appropriate mobile plan selections for the Megaline subscribers.

Both models showed promising results, with test accuracies above the required threshold of 0.75. However, the Random Forest (Adjusted) model demonstrated a slightly higher accuracy, suggesting that it may be a more suitable choice for plan recommendation in this scenario.

It is important to note that further analysis and evaluation should be conducted to fully assess the performance of these models. Additionally, exploring additional models and adjusting hyperparameters can provide a more comprehensive understanding of the available options and potentially yield even higher accuracies.

Overall, the Random Forest (Adjusted) model showcased superior performance with a higher test accuracy compared to the Gradient Boosting (Default) model. This highlights the importance of experimenting with different models and hyperparameters to select the most suitable model for accurately predicting mobile plan selections for Megaline subscribers.

Model 2: Gradient Boosting Classifier



In the test set, the model achieved an accuracy of 0.8087, indicating a strong performance on unseen data. This means that the model correctly predicted the plan (Smart or Ultra) for approximately 80.87% of the subscribers in the test set. The test accuracy is slightly higher than the validation accuracy, which suggests that the model's performance generalizes well to new, unseen data.

It's important to note that the test accuracy is above the threshold of 0.75 specified in the project instructions. This indicates that the model meets the desired level of accuracy and can be considered successful in recommending the appropriate plan for Megaline subscribers.

However, as mentioned earlier, it's recommended to assess additional evaluation metrics and conduct further analysis to gain a comprehensive understanding of the model's performance. This includes examining precision, recall, and F1 score, especially if the dataset is imbalanced or if certain types of errors are more critical.

Overall, with a test accuracy of 0.8087, the model demonstrates a solid ability to predict the appropriate plan for subscribers based on their behavior, indicating its effectiveness for Megaline's goal of recommending new plans to its subscribers.

# Sanity Check 

In [41]:
# Make predictions on the validation set
y_pred_val = model.predict(X_val)

# Create a DataFrame to compare predictions with actual values
sanity_check_df = pd.DataFrame({'Actual': y_val, 'Predicted': y_pred_val})

# Count the number of correct and incorrect predictions
correct_predictions = sanity_check_df[ sanity_check_df['Actual'] == sanity_check_df['Predicted'] ]
incorrect_predictions = sanity_check_df[ sanity_check_df['Actual'] != sanity_check_df['Predicted'] ]

# Calculate the accuracy of the model's predictions
sanity_check_accuracy = len(correct_predictions) / len(sanity_check_df)

print("Sanity Check Accuracy:", sanity_check_accuracy)
print("Number of Correct Predictions:", len(correct_predictions))
print("Number of Incorrect Predictions:", len(incorrect_predictions))

# Print a sample of the incorrect predictions
print("Sample of Incorrect Predictions:")
print(incorrect_predictions.head())


Sanity Check Accuracy: 0.7947122861586314
Number of Correct Predictions: 511
Number of Incorrect Predictions: 132
Sample of Incorrect Predictions:
      Actual  Predicted
2644       0          1
2767       1          0
3143       1          0
1669       1          0
2871       1          0


After performing the sanity check on the model, the following results were obtained:

Sanity Check Accuracy: The sanity check accuracy is 0.7947, which indicates that the model's predictions on the validation set match the actual values correctly approximately 79.47% of the time. This suggests that the model performs reasonably well in classifying subscribers into the appropriate plans based on their behavior.

Number of Correct Predictions: The model correctly predicted the plan (Smart or Ultra) for 511 subscribers in the validation set. These instances align with the actual values, demonstrating the model's ability to make accurate predictions.

Number of Incorrect Predictions: The model made 132 incorrect predictions on the validation set, where its predicted plan did not match the actual plan. These instances represent cases where the model struggled or misclassified subscribers based on their behavior.

Sample of Incorrect Predictions: The sample displays a subset of the incorrect predictions, showcasing specific instances where the model made errors. For example, in the first row, the model predicted a plan of 1 (Ultra), but the actual plan was 0 (Smart). Similarly, the model made incorrect predictions in the subsequent rows as well.

Analyzing the incorrect predictions can provide valuable insights into the limitations of the model and areas where it might struggle. It can help identify patterns, relationships, or particular features that may require further attention or feature engineering to improve the model's accuracy.

Overall, the sanity check results suggest that the model performs reasonably well, but there is room for improvement, particularly in reducing the number of incorrect predictions. These findings can inform further adjustments or refinements to the model and its underlying features to enhance its performance and accuracy.

Conclusion:

The "Predicting Mobile Plan Selection" project aimed to develop a data science model to analyze subscriber behavior and recommend appropriate plans for Megaline, a mobile carrier. By leveraging machine learning techniques and utilizing behavior data from subscribers who have already switched plans, the project sought to improve plan recommendations and increase customer satisfaction.

Throughout the project, we conducted data preprocessing and split the dataset into training, validation, and test sets. We explored different models and adjusted hyperparameters to achieve the highest possible accuracy. The model's performance was evaluated using validation and test datasets, with an accuracy threshold of 0.75 set as the project requirement.

The developed model demonstrated strong performance, surpassing the desired accuracy threshold. The validation accuracy of 0.7947 and test accuracy of 0.8087 indicated that the model effectively predicted the appropriate plan for subscribers based on their behavior. These results confirmed the model's ability to generalize well to unseen data and provided confidence in its effectiveness for plan recommendations.

Furthermore, a sanity check was performed to validate the model's predictions. The analysis of correct and incorrect predictions shed light on specific areas where the model may require further improvement or fine-tuning. By addressing these areas, the model can continue to evolve and enhance its accuracy over time.

In conclusion, the "Predicting Mobile Plan Selection" project offers Megaline a powerful tool for analyzing subscriber behavior and making personalized plan recommendations. By integrating the model into their systems, Megaline can enhance customer satisfaction, optimize plan adoption, and potentially increase revenue. Continuous monitoring and evaluation of the model's performance will ensure its effectiveness as subscriber behaviors and preferences evolve.

Ultimately, this project showcases the value of data science and machine learning in providing data-driven insights and driving informed decision-making in the telecommunications industry. With the developed model, Megaline is well-positioned to leverage subscriber behavior data and improve the overall customer experience.